text/0000-app-mixins.md

```diff
@@ -0,0 +1,299 @@
+
+A builder can use a stackpack by defining it in the `builder.toml` in the `[[stack.buildpacks]]` array of tables:
+
+```
+[[stack.buildpacks]]
+id = "<stackpack id>"
+uri = "<uri to stackpack>"
+```
+
+Each stackpack used by the builder will execute if it passes detection and it's provided mixins satisfy one of a project's required mixins.
+
+The stackpack's snapshot layer may be modified by writing a `launch.toml` file. The `[[processes]]` and `[[slices]]` tables may be used as normal, and a new `[[excludes]]` table will be introduced with the following schema:
+
+```
+[[excludes]]
+paths = ["<sub-path glob>"]
+cache = false
+```
+
+Where:
+
+* `paths` = a list of paths to exclude from the layer
+* `cache` = if true, the paths will be excluded from the launch image layer, but will be included in the cache layer.
+
+## Rebasing an App
+
+Before a launch image is rebased, the platform must re-run the any stackpacks that were used to build the launch image against the new run-image. Then, the rebase operation can be performed as normal, while including the stackpack layers as part of the stack. This will be made possible by including the stackpack in the run-image, but because the stackpack detect phase is not run, the operation does not need access to the application source.
+
+## Example: Apt Buildpack
+
+A buildpack that can install an arbitrary list of mixins would have a `buildpack.toml` like this:
+
+ ```toml
+[buildpack]
+id = "example/apt"
+privileged = true
+mixins = ["[^=]+"]
+
+[[stacks]]
+id = "io.buildpacks.stacks.bionic"
+```
+
+Its `bin/detect` would have the following contents:
+
+ ```bash
+#!/usr/bin/env bash
+
+exit 0
+```
+
+Its `bin/build` would have the following contents:
+
+ ```bash
+#!/usr/bin/env bash
+
+apt update -y
+
+for package in $(cat $3 | yj -t | jq -r ".entries | .[] | .name"); do
+  apt install $package
+done
+
+cat << EOF > launch.toml
+[[excludes]]
+paths = [ "/var/cache" ]
+cache = true
+EOF
+```
+
+## Example: CA Certificate Buildpack
+
+Support for custom CA Certificates can be accomplished with two buildpacks: a stackpack that can install the cert, and a normal buildpack that can provide a cert in the build plan.
+
+### CA Cert Installer Stackpack
+
+A buildpack that works on the `ubuntu-20` stack, which defines a `type=` prefix for mixins, and installs custom CA Certificates would have a `buildpack.toml` that looks like this:
+
+```toml
+[buildpack]
+id = "example/cacerts"
+privileged = true
+mixins = ["cacert"]
+
+[[stacks]]
+id = "ubuntu-20"
+```
+
+Its `bin/detect` would have the following contents:
+
+ ```bash
+#!/usr/bin/env bash
+
+cat << EOF > $2
+[[provides]]
```

Comment:
I support `/bin/detect` for stackpacks, so that they can validate that the stack image itself is in a valid state. However, I'm reluctant to allow mixins to be specified dynamically during detect, given that the app directory is not available.

Reply:
@sclevine what if it's not based on app source, but some other input from the platform (like a list of mixins that are needed)

Reply:
Addressed in 8ee98d430e2ce55652cd53ae5ed9f344bd888889

Reply:
> However, I'm reluctant to allow mixins to be specified dynamically during detect, given that the app directory is not available.

I agree that the stackpacks shouldn't dynamically decide which mixins to provide during detect, but regular buildpacks (which have access to the app dir) should probably be able to require specific mixins during detect.